### ISSUE TITLE
Memory Leak in PyTorch 1.10.1

### ISSUE THREAD

**Issue by trzy** [pytorch/pytorch#71495]
### 🐛 Describe the bug

PyTorch is leaking CUDA memory. I have a model (Faster R-CNN) that produces a tensor called 'proposals' that is computed from the output of a model stage and then detached from the graph before being used as input into another stage. Both of these stages have trainable weights but the key is that 'proposals' should be treated as a constant.

I've created a gist: https://gist.github.com/trzy/8eb1452665248b91064d0efbfc35dde9
Sorry for the size but it absolutely requires valid data to be bundled in to work. You should see memory grow without bounds when running this. But, search for "**** UNCOMMENT THE LINE BELOW" in the code, uncomment said line, and watch the memory leak disappear. 

More info:

Despite detaching the proposals in FasterRCNNModel.train_step(), tensor, memory usage grows, and using a CUDA memory profiler, I pinpointed the problem to a couple of intermediate tensors (in t_convert_deltas_to_boxes() in the sample gist) not being released. However, nothing should be keeping these alive. The full gist demonstrates the problem but at a high level, here is what is happening:

In FasterRCNNModel.train_step(), you can see proposals and some ground truth data are generated *during training* but then detached before any further use:

```
    # Stage 1: Extract features
    feature_map = self._stage1_feature_extractor(image_data = image_data)

    # Stage 2: Generate object proposals using RPN 
    rpn_score_map, rpn_box_deltas_map, proposals = self._stage2_region_proposal_network(
      feature_map = feature_map,
      image_shape = image_shape,  # each image in batch has identical shape: (num_channels, height, width)
      anchor_map = anchor_map,
      anchor_valid_map = anchor_valid_map,
      max_proposals_pre_nms = 12000,
      max_proposals_post_nms = 2000
    )

    # Sample random mini-batch of anchors (for RPN training)
    gt_rpn_minibatch_map = self._sample_rpn_minibatch(
      rpn_map = gt_rpn_map,
      object_indices = gt_rpn_object_indices,
      background_indices = gt_rpn_background_indices
    )

    # Assign labels to proposals and take random sample (for detector training)
    proposals, gt_classes, gt_box_deltas = self._label_proposals(
      proposals = proposals,
      gt_boxes = gt_boxes[0], # for now, batch size of 1
      min_background_iou_threshold = 0.0,
      min_object_iou_threshold = 0.5
    )
    proposals, gt_classes, gt_box_deltas = self._sample_proposals(
      proposals = proposals,
      gt_classes = gt_classes,
      gt_box_deltas = gt_box_deltas,
      max_proposals = self._proposal_batch_size,
      positive_fraction = 0.25
    )

    # Make sure RoI proposals and ground truths are detached from computational
    # graph so that gradients are not propagated through them. They are treated
    # as constant inputs into the detector stage.
    proposals = proposals.detach()
    gt_classes = gt_classes.detach()
    gt_box_deltas = gt_box_deltas.detach()

    # Stage 3: Detector
    detector_classes, detector_box_deltas = self._stage3_detector_network(
      feature_map = feature_map,
      proposals = proposals
    )

    # Compute losses
    ...
```
train_step returns a single float (.item() from one of the losses), therefore there is no danger of the graph being persisted between train steps. Note that proposals, gt_classes, and gt_box_deltas are detached and in between them being generated and detached are not used anywhere. 

proposals is derived from rpn_box_deltas_map but neither rpn_score_map nor rpn_box_deltas_map depend on proposals. That is, the two maps are the outputs of a neural network and then proposals is computed from one of them. 

The RPN stage is where I have to insert the fix (which makes no sense):

```
def forward(self, feature_map, image_shape, anchor_map, anchor_valid_map, max_proposals_pre_nms, max_proposals_post_nms):
    # Pass through the network
    y = F.relu(self._rpn_conv1(feature_map))
    objectness_score_map = t.sigmoid(self._rpn_class(y))
    box_deltas_map = self._rpn_boxes(y)

    # Transpose shapes to be more convenient:
    #   objectness_score_map -> (batch_size, height, width, num_anchors)
    #   box_deltas_map       -> (batch_size, height, width, num_anchors * 4)
    objectness_score_map = objectness_score_map.permute(0, 2, 3, 1).contiguous()
    box_deltas_map = box_deltas_map.permute(0, 2, 3, 1).contiguous()

    # Returning to CPU land by extracting proposals as lists (NumPy arrays)
    anchors, objectness_scores, box_deltas = self._extract_valid(
      anchor_map = anchor_map,
      anchor_valid_map = anchor_valid_map,
      objectness_score_map = objectness_score_map,
      box_deltas_map = box_deltas_map
    )

    # **** UNCOMMENT THE LINE BELOW TO "FIX" MEMORY LEAK ****
    # Detach from graph to avoid backprop. According to my understanding, this
    # should be redundant here because we later take care to detach the
    # proposals (in FasterRCNNModel). However, there is a memory leak involving
    # t_convert_deltas_to_boxes() if this is not done here. Ultimately, the
    # numerical results are not affected. Proposals returned from this function
    # are supposed to be constant and are fed into the detector stage. See any
    # commit prior to 209141c for an earlier version of the code here that
    # performed all operations on CPU using NumPy, which was slightly slower
    # but equivalent.
    #box_deltas = box_deltas.detach()

    # Convert regressions to box corners
    proposals = math_utils.t_convert_deltas_to_boxes(
      box_deltas = box_deltas,
      anchors = t.from_numpy(anchors).cuda(),
      box_delta_means = t.tensor([0, 0, 0, 0], dtype = t.float32, device = "cuda"),
      box_delta_stds = t.tensor([1, 1, 1, 1], dtype = t.float32, device = "cuda")
    )

    # Keep only the top-N scores. Note that we do not care whether the
    # proposals were labeled as objects (score > 0.5) and peform a simple
    # ranking among all of them. Restricting them has a strong adverse impact
    # on training performance.
    sorted_indices = t.argsort(objectness_scores)                   # sort in ascending order of objectness score
    sorted_indices = sorted_indices.flip(dims = (0,))               # descending order of score
    proposals = proposals[sorted_indices][0:max_proposals_pre_nms]  # grab the top-N best proposals
    objectness_scores = objectness_scores[sorted_indices][0:max_proposals_pre_nms]  # corresponding scores

    # Clip to image boundaries
    proposals[:,0:2] = t.clamp(proposals[:,0:2], min = 0)
    proposals[:,2] = t.clamp(proposals[:,2], max = image_shape[1])
    proposals[:,3] = t.clamp(proposals[:,3], max = image_shape[2])

    # Remove anything less than 16 pixels on a side
    height = proposals[:,2] - proposals[:,0]
    width = proposals[:,3] - proposals[:,1]
    idxs = t.where((height >= 16) & (width >= 16))[0]
    proposals = proposals[idxs]
    objectness_scores = objectness_scores[idxs]

    # Perform NMS
    idxs = nms(
      boxes = proposals,
      scores = objectness_scores,
      iou_threshold = 0.7
    )
    idxs = idxs[0:max_proposals_post_nms]
    proposals = proposals[idxs]

    # Return network outputs as PyTorch tensors and extracted object proposals
    return objectness_score_map, box_deltas_map, proposals
```

Note the detach statement and note what is returned. Removing said detach statement (as in the gist I've supplied) produces a leak. I *think* the memory leak is inside of t_convert_deltas_to_boxes(), which is pretty innocuous:

```
def t_convert_deltas_to_boxes(box_deltas, anchors, box_delta_means, box_delta_stds):
  box_deltas = box_deltas * box_delta_stds + box_delta_means
  center = anchors[:,2:4] * box_deltas[:,0:2] + anchors[:,0:2]  # center_x = anchor_width * tx + anchor_center_x, center_y = anchor_height * ty + anchor_center_y
  size = anchors[:,2:4] * t.exp(box_deltas[:,2:4])              # width = anchor_width * exp(tw), height = anchor_height * exp(th)
  boxes = t.empty(box_deltas.shape, dtype = t.float32, device = "cuda")
  boxes[:,0:2] = center - 0.5 * size                            # y1, x1
  boxes[:,2:4] = center + 0.5 * size                            # y2, x2
  return boxes
```

### Versions

Collecting environment information...
PyTorch version: 1.10.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.13.0-25-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 3090
Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.22.1
[pip3] torch==1.10.1+cu113
[pip3] torchaudio==0.10.1+cu113
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.4
[pip3] torchvision==0.11.2+cu113
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1

**Comment by twsl** [comment]
I have a similar issue with FasterRCNN from the torch vision lib. Mind sharing how you were able to detect which objects remained in memory?

**Comment by trzy** [comment]
I used a CUDA memory profiler I found on GitHub. My Faster R-CNN repo is now public (https://github.com/trzy/FasterRCNN) and the profiler is in pytorch/FasterRCNN/profile.py. See line 96 of pytorch/FasterRCNN/__main__.py for usage. It's quite simple.

Although the output seems overwhelming at first, I quickly found that I could actually track it by hand and discovered that there were two tensors that where never being freed (for every '+' line there should be a corresponding '-' line). But while I was able to identify roughly where these tensors were (temporary tensors used in the construction of an output tensor), I still do not understand why they cause a leak.

I hope someone on the PyTorch team downloads my example app because it illustrates the problem. In my repo, you can see the problem and the solution fairly clearly by examining line 114 of pytorch/FasterRCNN/models/rpn.py, where I detach a tensor despite there being no obvious reason to do so because on line 301 of pytorch/FasterRCNN/models/faster_rcnn.py, I detach the only descendants of that tensor. If I'm wrong here, I'd love to know why, but for now, I do believe this is a bug and it's interesting to hear that it is also present in their official Faster R-CNN code (which I didn't look at when I wrote mine).
